### PR TITLE
merging diffractive_imaging into dev to coincide with DGP paper

### PR DESCRIPTION
**Pull request by arthurmccray**
This PR brings some new features, bugfixes, and general improvements. There are a couple of changes outside `diffractive_imaging`, with the only semi-significant one being a bugfix and simplification in `autoserialize`. 

Example notebooks can be found at [quantem-tutorials/ml](https://github.com/electronmicroscopy/quantem-tutorials/tree/ml) 

Still to do -- create an example notebook showing the direct ptycho -> iterative ptycho workflow (and potentially smooth that over with some new methods)

### CODE REVIEW

File: src/quantem/core/io/serialize.py
Comment (arthurmccray) [inline src/quantem/core/io/serialize.py]: Previously I was having issues where autoreload was causing autoserializable subclasses to not be recognized (this is due to `isinstance` failing with autoreload, which seems to be a known issue). I had fixed it, but somehow that caused a bug where optimizer states were not being re-loaded about 50% of the time. I have no idea what caused this, but rerunning the same cell of a notebook repeatedly did not give the same results every time. 

This is (I think) a fix to both issues. I've tested it and it seems pretty robust so far, but time will tell...

File: src/quantem/diffractive_imaging/logger_ptychography.py
Comment (arthurmccray) [inline src/quantem/diffractive_imaging/logger_ptychography.py]: epoch -> iter

File: src/quantem/core/ml/optimizer_mixin.py [hunk 1]
```diff
@@ -106,6 +113,7 @@ def set_optimizer(self, opt_params: dict | None = None) -> None:
 
         if isinstance(opt_type, type):
             self._optimizer = opt_type(params, **opt_params)
+            self.optimizer_params["type"] = opt_type.__module__.split(".")[-1]
```
Comment (gvarnavi) [inline src/quantem/core/ml/optimizer_mixin.py]: this seems dangerous / hacky.. is there a more robust way to achieve what you're trying to do? where is this used?

File: src/quantem/core/utils/diffractive_imaging_utils.py
Comment (gvarnavi) [inline src/quantem/core/utils/diffractive_imaging_utils.py]: These should probably move to a more general utils file, like we need to fit a probe circle for all non-calibrated datasets

File: src/quantem/diffractive_imaging/dataset_models.py [hunk 2]
```diff
@@ -63,6 +64,8 @@ def __init__(
         self.verbose = verbose
         self._preprocessed = False
         self._preprocessing_params = {}  # for serialization and reloading
+        self._com_rotation_rad = 0  # default
```
[the remark this answers is not in the document]
Comment (gvarnavi) [reply src/quantem/diffractive_imaging/dataset_models.py]: I think Steph wants us to switch to specifying everything in degrees, but ok for now to be consistent with direct ptycho

File: src/quantem/diffractive_imaging/dataset_models.py [hunk 3]
```diff
@@ -164,6 +169,23 @@ def positions_px_fractional(self) -> torch.Tensor:
     # endregion --- optimizable parameters ---
 
     # region --- buffers ---
+    @property
```
Comment (gvarnavi) [inline src/quantem/diffractive_imaging/dataset_models.py]: I suspect this is used elsewhere to mask updates?

File: src/quantem/diffractive_imaging/ptychography_base.py [hunk 4]
```diff
@@ -476,12 +520,15 @@ def probe_model(self, model: ProbeModelType | type):
...
-        )
+        try:
+            self._probe_model.set_initial_probe(
+                self.roi_shape,
+                self.reciprocal_sampling,
+                self.dset.mean_diffraction_intensity,
+                device=self.device,
+            )
+        except Exception:
+            pass
```
Comment (gvarnavi) [inline src/quantem/diffractive_imaging/ptychography_base.py]: why is this allowed to fail silently again?

File: src/quantem/diffractive_imaging/ptycho_utils.py
[the remark this answers is not in the document]
Comment (arthurmccray) [reply src/quantem/diffractive_imaging/ptycho_utils.py]: I added a `val_ratio` attribute (intended to be set during `ptycho.preprocess`) that allows for designating some fraction of the scan positions as validation data, so the loss will be recorded but they will not be used to train. This is potentially useful for visualizing overfitting to noise, though in practice you'll of course want to run your final recons learning from 100% of the scan positions. 

And to confirm, nothing is changed when `val_ratio` is left at the default value of `0`.

File: src/quantem/diffractive_imaging/ptychography_base.py (hunk 4 above)
Comment (arthurmccray) [reply src/quantem/diffractive_imaging/ptychography_base.py]: This only fails if the dset hasn't been preprocessed, and is then called in ptycho.preprocess instead. I'm updating it to not be an explicit if/else call rather than try/except with a comment to explain.

File: src/quantem/core/datastructures/dataset4d.py
[the remark this answers is not in the document]
Comment (arthurmccray) [reply src/quantem/core/datastructures/dataset4d.py]: Seems reasonable to me. I don't think there's a way to make it play nicely with static type checking, at least not easily.

**Review by gvarnavi — APPROVED** [review]
